src/Native/Runtime/gcrhenv.cpp

```diff
@@ -1504,6 +1416,159 @@ bool GCToEEInterface::EagerFinalized(Object* obj)
+        // Redhawk thread state which requires running in the new thread's context.
+        auto threadStub = [](void* argument) -> DWORD
+        {
+            SuspendableThreadStubArguments* pStartContext = (SuspendableThreadStubArguments*)argument;
+
+            // Initialize the Thread for this thread. The false being passed indicates that the thread store lock
+            // should not be acquired as part of this operation. This is necessary because this thread is created in
+            // the context of a garbage collection and the lock is already held by the GC.
+            ASSERT(GCHeapUtilities::IsGCInProgress());
+            ThreadStore::AttachCurrentThread(false);
+
+            Thread* pThread = ThreadStore::GetCurrentThread();
+            pThread->SetGCSpecial(true);
+
+            // Inform the GC which Thread* we are.
+            pStartContext->m_pThread = pThread;
+
+            auto realStartRoutine = pStartContext->m_pRealStartRoutine;
+            void* realContext = pStartContext->m_pRealContext;
+
+            pStartContext->m_ThreadStartedEvent.Set();
+
+            STRESS_LOG_RESERVE_MEM(GC_STRESSLOG_MULTIPLY);
+
+            realStartRoutine(realContext);
+
+            return 0;
+        };
+
+        if (!PalStartBackgroundGCThread(threadStub, &threadStubArgs))
+        {
+            threadStubArgs.m_ThreadStartedEvent.CloseEvent();
+            return NULL;
+        }
+
+        uint32_t res = threadStubArgs.m_ThreadStartedEvent.Wait(INFINITE, FALSE);
+        threadStubArgs.m_ThreadStartedEvent.CloseEvent();
+        ASSERT(res == WAIT_OBJECT_0);
+
+        ASSERT(threadStubArgs.m_pThread != NULL);
+        return threadStubArgs.m_pThread != NULL;
+    }
+
+    bool CreateNonSuspendableThread(
+        void (*threadStart)(void*),
+        void* argument,
+        const char* name)
+    {
+        ThreadStubArguments args;
+        args.Argument = argument;
+        args.ThreadStart = threadStart;
+        args.Thread = INVALID_HANDLE_VALUE;
+        if (!args.ThreadStartedEvent.CreateAutoEventNoThrow(FALSE))
+        {
+            return false;
+        }
+
+        auto threadStub = [](void* argument) -> DWORD
+        {
+            ThreadStubArguments* args = static_cast<ThreadStubArguments*>(argument);
+            assert(args != nullptr);
+
+            ClrFlsSetThreadType(ThreadType_GC);
+            STRESS_LOG_RESERVE_MEM(GC_STRESSLOG_MULTIPLY);
+
+            args->HasStarted = true;
+            auto threadStart = args->ThreadStart;
+            void* threadArgument = args->Argument;
+            args->ThreadStartedEvent.Set();
+
+            // The stub args cannot be used once the event is set, since that releases wait on the
+            // event in the function that created this thread and the stubArgs go out of scope.
+            threadStart(threadArgument);
+            return 0;
+        };
+
+        args.Thread = Thread::CreateUtilityThread(Thread::StackSize_Medium, threadStub, &args);
```

Comment:
I've copied `CreateNonSuspendableThread` from coreclr. There's no `Thread::CreateUtilityThread` in corert. I was poking a bit at `PalStartBackgroundWork` but it's unix implementation does not return a handle, not sure if this is the right direction.
`CreateUtilityThread` calls `CreateThread` directly, how does that work cross platform and what's the intention behind the `FEATURE_PAL` define?

Reply:
The Unix impl should be pretty close to the existing `GCToOSInterface::CreateThread`. The main difference should be that suspendable threads get an extra `TreadStore::AttachCurrentThread` call.